File: 2023/03.16/조수연/단어_변환.java
import java.util.*;

public class Problem {
    public static void main(String[] args) {
        new SolutionJava().solution("hit","cog", new String[]{"hot", "dot", "dog", "lot", "log", "cog"});
    }
}

class WordInfo{
    String word;
    ArrayList indexList = new ArrayList<Integer>();

    public WordInfo(String word, ArrayList indexList) {
        this.word = word;
        this.indexList = indexList;
    }
}
class SolutionJava {
    public int solution(String begin, String target, String[] words) {
        int answer = 0;
        Queue<WordInfo> queue = new LinkedList<>();

        boolean isExist = false;

        for (int i = 0; i < words.length; i++) { // 단어가 존재하는지 체크
            if (words[i].equals(target)) {
                isExist = true;
                break;
            }
        }

        if (isExist){
            for (int i = 0; i < words.length; i++) { // 글자 수 하나 다른 단어 선택하기
                int count = 0;
                for (int j = 0; j < begin.length(); j++) {
                    if (begin.charAt(j) != words[i].charAt(j)) count++;
                    if (count > 1) break;
                }

                if (count == 1){
                    ArrayList<Integer> indexList = new ArrayList<>();
                    indexList.add(i);
                    queue.add(new WordInfo(words[i],indexList));
                }

            }
            answer = bfs(queue,target,words);
        }

        return answer;
    }

    public int bfs(Queue<WordInfo> queue, String target, String[] words){
        int count = 1;

        while (!queue.isEmpty()){
            for (int i = 0; i < queue.size(); i++) {
                WordInfo info = queue.poll();
                String changeWord = info.word; // 바뀐 단어
                ArrayList<Integer> indexList = info.indexList; // 방문한 인덱스 리스트

                if (changeWord.equals(target)) return count; // poll 했을 때 target이면 count 반환

                for (int j = 0; j < words.length; j++) {
                    if (!indexList.contains(j)){ // 중복 제거
                        int diffCnt = 0;
                        for (int k = 0; k < changeWord.length(); k++) {
                            if (changeWord.charAt(k) != words[j].charAt(k)) diffCnt++;
                            if (diffCnt > 1) break;
                        }

                        if (diffCnt == 1){ // 문자가 하나만 다르면 queue에 add
                            indexList.add(j);
                            queue.add(new WordInfo(words[j],indexList));
                        }
                    }
                }
            }
            count++;
        }
        return count;
    }
}

File: 2023/03.16/조수연/단어_변환.kt
import java.util.LinkedList
import java.util.Queue

fun main() {
    Solution().solution(begin = "hit", target = "cog", words = arrayOf("hot", "dot", "dog", "lot", "log"))
}

class Solution {
    fun solution(begin: String, target: String, words: Array<String>): Int {
        var answer = 0
        val queue: Queue<Pair<String, ArrayList<Int>>> = LinkedList()
        if (words.contains(target)) { // 글자 수 하나 다른 단어 선택하기
            for (i in words.indices) {
                var count = 0
                for (j in begin.indices) {
                    if (begin[j] != words[i][j]) count++
                    if (count > 1) break
                }
                if (count == 1) queue.add(Pair(words[i], arrayListOf(i)))
            }
            answer = bfs(queue, target, words) // 탐색 시작
        }
        return answer
    }

    fun bfs(queue: Queue<Pair<String, ArrayList<Int>>>, target: String, words: Array<String>): Int {
        var count = 1

        while (queue.isNotEmpty()) {
            for (i in queue.indices) {
                val info = queue.poll()
                val changeWord = info.first // 바뀐 단어
                val indexList = info.second // 방문한 인덱스 리스트

                if (changeWord == target) return count // poll 했을 때 target이면 count 반환

                for (j in words.indices) {
                    if (!indexList.contains(j)) { // 중복 제거
                        var count = 0
                        for (k in changeWord.indices) {
                            if (changeWord[k] != words[j][k]) count++
                            if (count > 1) break
                        }

                        if (count == 1) { // 문자가 하나만 다르면 queue에 add
                            indexList.add(j)
                            queue.add(Pair(words[j], indexList))
                        }
                    }
                }
            }
            count++
        }
        return count
    }
}